### PR TITLE
Customize rotation averaging cost

### PR DESCRIPTION
**Pull request by FangLinHe**

### CODE REVIEW

File: src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h [hunk 1]
```diff
@@ -57,7 +64,56 @@ class NonlinearRotationEstimator : public RotationEstimator {
...
+        FindOrNull(*global_orientations, view_id_pair.second);
+
+    // Do not add the relative rotation constaint if it requires an orientation
+    // that we do not have an initialization for.
+    if (rotation1 == nullptr || rotation2 == nullptr) {
+      continue;
+    }
+
+    ceres::CostFunction* cost_function =
+        CostFunctionGenerator::Create(view_pair.second.rotation_2);
```
Comment (billyzs) [inline src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h]: doesn't `PairwiseRotationError::Create` need a second double param `weight`?

File: src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h [hunk 2]
```diff
@@ -47,6 +48,12 @@ namespace theia {
 // Computes the global rotations given relative rotations and an initial guess
 // for the global orientations. Nonlinear optimization is performed with Ceres
 // using a SoftL1 loss function to be robust to outliers.
+//
+// CostFunctionGenerator is a class which defines a function `Create` that
+// generates a ceres::CostFunction* given a relative rotation in vector 3D:
+// ceres::CostFunction* Create(const Eigen::Vector3d& relative_rotation)
+// Check PairwiseRotationError as an example.
+template <class CostFunctionGenerator=PairwiseRotationError>
```
Comment (billyzs) [inline src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h]: I think in general this does what we want. Template classes in C++ are a bit hard to wrangle though, e.g. we have to do what you did here which is putting everying in the header and that adds to code size and compile speed. I proposed another soln which uses `std::function`, feel free to take a look and decide which feels more natural.

File: src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h (hunk 1 above)
Comment (billyzs) [reply src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h]: oh I see, you made it a default param

File: src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h (hunk 2 above)
Comment (FangLinHe) [reply src/theia/sfm/global_pose_estimation/nonlinear_rotation_estimator.h]: Thanks for the suggestion. I had a look at your changes and it looks good to me except the fact that the with `std::function`, it's not flexible to have different parameters for different cost functions, which happens very likely.

It's actually possible to separate the template into header and source file, but in source file it looks quite ugly, and I think we also need to specify the templates explicitly in source file.

**Pull request closed without merging**